module/zfs/zio.c

```diff
@@ -3083,15 +3084,16 @@ zio_write_gang_block(zio_t *pio, metaslab_class_t *mc)
 	boolean_t has_data = !(pio->io_flags & ZIO_FLAG_NODATA);
 
 	/*
-	 * If one copy was requested, store 2 copies of the GBH, so that we
-	 * can still traverse all the data (e.g. to free or scrub) even if a
-	 * block is damaged.  Note that we can't store 3 copies of the GBH in
-	 * all cases, e.g. with encryption, which uses DVA[2] for the IV+salt.
+	 * Store multiple copies of the GBH, so that we can still traverse
+	 * all the data (e.g. to free or scrub) even if a block is damaged.
+	 * This value respects the redundant_metadata property. Note that
+	 * we can't store 3 copies of the GBH in all cases, e.g. with
+	 * encryption, which uses DVA[2] for the IV+salt.
 	 */
-	int gbh_copies = copies;
-	if (gbh_copies == 1) {
-		gbh_copies = MIN(2, spa_max_replication(spa));
-	}
+	int gbh_copies = MIN(gio->io_prop.zp_gang_copies,
+	    MAX(spa_max_replication(spa) - 1, copies));
```

Comment:
Do we need any logic here after we properly set `zp_gang_copies` above?

Reply:
We need to fix the max number of gang copies to be 2 if we have encryption, but we should do that in dmu_write_policy as well.